gcsfs/core.py

```diff
@@ -271,14 +267,36 @@ def __init__(
         self._endpoint = endpoint_url
         self.session_kwargs = session_kwargs or {}
 
-        self.credentials = GoogleCredentials(project, access, token, check_connection)
+        if check_connection:
+            warnings.warn(
+                "The `check_connection` argument is deprecated and will be removed in a future release.",
+                DeprecationWarning,
+            )
+
+        if token is None:
+            self._guess_credentials(project, access)
+        else:
+            self.credentials = GoogleCredentials(project, access, token)
 
         if not self.asynchronous:
             self._session = sync(
                 self.loop, get_client, timeout=self.timeout, **self.session_kwargs
             )
             weakref.finalize(self, self.close_session, self.loop, self._session)
 
+    def _guess_credentials(self, project, access):
+        for meth in ["google_default", "cache", "cloud", "anon"]:
+            try:
+                self.credentials = GoogleCredentials(project, access, meth)
+                logger.debug("Connected with method %s", meth)
+                break
+            except google.auth.exceptions.GoogleAuthError as e:
+                # GoogleAuthError is the base class for all authentication
+                # errors
+                logger.debug('Connection with method "%s" failed' % meth, exc_info=e)
+        else:
+            raise RuntimeError("All connection methods have failed!")
```

Comment:
anon never fails

Reply:
thanks, that's good to know. i think it's still a good idea to raise instead of leaving `self.credentials=None`, in case something unexpected happens. but I can add a comment, indicating that the else branch should technically never be executed.

Reply:
ok